tests/Unit/HtmlProcessor/CssVariableEvaluatorTest.php

```diff
@@ -0,0 +1,394 @@
+        $cssInliner->inlineCss($css);
+
+        return CssVariableEvaluator::fromDomDocument($cssInliner->getDomDocument());
+    }
+
+    /**
+     * @return array<non-empty-string, array{css: non-empty-string, expect: non-empty-string}>
+     */
+    public function provideCssUsingVaraiblesAndExpectedHtmlFragmentAfterInliningAndEvaluation(): array
+    {
+        return [
+            'undefined variable' => [
+                'css' => '
+                    p {
+                        color: var(--text-color);
+                    }
+                ',
+                'expect' => '<p style="color: var(--text-color);">',
+            ],
+            'variable defined in root, used in descendant' => [
+                'css' => '
+                    :root {
+                        --text-color: green;
+                    }
+                    p {
+                        color: var(--text-color);
+                    }
+                ',
+                'expect' => '<p style="color: green;">',
+            ],
+            'variable defined in parent, used in child' => [
+                'css' => '
+                    p {
+                        --text-color: green;
+                    }
+                    code {
+                        color: var(--text-color);
+                    }
+                ',
+                'expect' => '<code style="color: green;">',
+            ],
+            'variable defined and used in same element' => [
+                'css' => '
+                    p {
+                        --text-color: green;
+                    }
+                    p {
+                        color: var(--text-color);
+                    }
+                ',
+                // The variable definition is not removed, but its value should nonetheless be applied
+                'expect' => '<p style="--text-color: green; color: green;">',
+            ],
+            'variable defined in parent and root' => [
+                'css' => '
+                    :root {
+                        --text-color: red;
+                    }
+                    p {
+                        --text-color: green;
+                    }
+                    code {
+                        color: var(--text-color);
+                    }
+                ',
+                'expect' => '<code style="color: green;">',
+            ],
+            'variable defined and used in same element, also defined in root' => [
+                'css' => '
+                    :root {
+                        --text-color: red;
+                    }
+                    p {
+                        --text-color: green;
+                    }
+                    p {
+                        color: var(--text-color);
+                    }
+                ',
+                // The variable definition is not removed, but its value should nonetheless be applied
+                'expect' => '<p style="--text-color: green; color: green;">',
```

Comment:
This example tends to boggle my mind … is this the way CSS custom properties are expected to work?

Reply:
Yes, the value comes from the a definition for the element itself or the closest ancestor.  Or, in other words, custom properties are inherited.

Reply:
The `@property` rule allows changing of inheritance behaviour.  It would be possible to support it (somehow), but I don't see any great need at the moment.  I don't think it is commonly used (I wasn't even aware of it) - whereas CSS variables are everywhere nowadays (hence my keenness to support them).